include/clap/ext/draft/context-menu.h

```diff
@@ -0,0 +1,121 @@
+{
+   uint32_t kind;
+   clap_id id;
+} clap_context_menu_target_t;
+
+// Flags for the context menu accelerator
+enum {
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_SHIFT = (1 << 0),
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_CONTROL = (1 << 1),
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_ALT = (1 << 2),
+
+   // Windows key or Command key
+   CLAP_CONTEXT_MENU_ACCELERATOR_MODIFIER_META1 = (1 << 3),
+};
+
+// Describes a single key strike with its modifiers
+typedef struct clap_context_menu_accelerator_key {
+   uint32_t flags;
+   uint32_t key; // ASCII key
+} clap_context_menu_accelerator_key_t;
+
+// Describes an accelerator which can consist of a key strike combo
+typedef struct clap_context_menu_accelerator {
+   uint32_t combo_size;
+   clap_context_menu_accelerator_key_t keys[4];
+} clap_context_menu_accelerator_t;
+
+// Context menu builder
+typedef struct clap_context_menu_builder {
+   void *ctx;
+
+   // accelerator may be null
+   bool(CLAP_ABI *add_entry)(const struct clap_context_menu_builder *builder,
+                             const char                             *label,
+                             const clap_context_menu_accelerator_t  *accelerator,
+                             bool                                    is_enabled,
+                             clap_id                                 action_id);
+
+   // accelerator may be null
+   bool(CLAP_ABI *add_checkmark)(const struct clap_context_menu_builder *builder,
+                                 const char                             *label,
+                                 const clap_context_menu_accelerator_t  *accelerator,
+                                 bool                                    is_enabled,
+                                 bool                                    is_checked,
+                                 clap_id                                 action_id);
+
+   bool(CLAP_ABI *add_separator)(const struct clap_context_menu_builder *builder);
+
+   // Each begin_submenu() must have a corresponding end_submenu()
+   bool(CLAP_ABI *begin_submenu)(const struct clap_context_menu_builder *builder,
+                                 const char                             *label);
+
+   bool(CLAP_ABI *end_submenu)(const struct clap_context_menu_builder *builder);
+} clap_context_menu_builder_t;
+
+typedef struct clap_plugin_context_menu {
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *populate)(const clap_plugin_t               *plugin,
+                            const clap_context_menu_target_t  *target,
+                            const clap_context_menu_builder_t *builder);
+
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *perform)(const clap_plugin_t              *plugin,
+                           const clap_context_menu_target_t *target,
+                           clap_id                           action_id);
+} clap_plugin_context_menu_t;
```

Comment:
To make sure I understand things right here, can clap_plugin_context_menu->populate() be called from the host on the plugin to query for menu items that the host can add to its plugin specific context menu that is shown in the host device panel? (without the plugin showing or having its own GUI)

And, I assume all of this API is intended to be [main-thread] only? It'd probably be good to state that somewhere.

Reply:
yes, and for the menu builder I think it is implied, but I can write it explicitly.